Bug fixes for "visible groups"

cooperka opened the pull request:
### Changes

#### Hide groups from repeat picker

Fixes a UI bug I found in the recently-merged https://github.com/opendatakit/collect/pull/2764. If a form had a visible group on the same screen as a repeat group, the visible group showed up inside the picker.

This simply adds a guard to make sure nothing shows up in a repeat picker except for repeat group instances.

Demo form: [repeat_group_form.xml](https://github.com/opendatakit/collect/files/2789289/repeat_group_form.xml.txt)

#### Fix ANR when navigating up to initial page

Fixes https://github.com/opendatakit/collect/issues/2824

Previously it would try to step **back** once more, but `refreshView` steps **forward** automatically from the beginning of the form, which led to an infinite loop.

See issue for verification steps.

#### Fix bin icon not showing up for some repeat instances

Fixes https://github.com/opendatakit/collect/issues/2821

Now uses the same logic as `FormEntryActivity` to detect repeats. A regression caused the bin icon to be hidden for repeat instances nested inside groups, because the groups themselves are skipped over in the hierarchy view.

See issue for verification steps.

#### Remove redundant bottom "go up" button

Resolves https://github.com/opendatakit/collect/issues/2831

cooperka committed:
Also hide groups from repeat picker

grzesiek2010 commented:
Looks as if this fixes the bug but I don't get your description:
>If a form had a visible group on the same screen as a repeat group, the visible group showed up inside the picker.

They are not on the same screen... Am I missing anything?

Btw. I discovered another bug: https://github.com/opendatakit/collect/issues/2824

cooperka commented:
> They are not on the same screen... Am I missing anything?

By "same screen" I mean the visible group (called Group Name) is on the same level of the hierarchy as both repeat groups (Grp1 and Grp2).

<img width="290" alt="screen shot 2019-01-24 at 20 14 38" src="https://user-images.githubusercontent.com/2047062/51718758-bd5e0100-2014-11e9-8ddc-ebb08901be6e.png">

When you enter the repeat picker for either repeat, the visible group shows up inside the picker.

<img width="292" alt="screen shot 2019-01-24 at 20 14 57" src="https://user-images.githubusercontent.com/2047062/51718761-c18a1e80-2014-11e9-85d7-c108b817e8c5.png">

And thank you for the other bug report, I'll look into it.

cooperka committed:
Fix ANR when navigating up to initial page

Previously it would try to step *back* once more, but
refreshView steps *forward* automatically from the beginning
of the form, which led to an infinite loop.

cooperka committed:
Fix bin icon not showing up in some forms

Now uses the same logic as FormEntryActivity to detect repeats.

mmarciniak90 commented:
@cooperka @grzesiek2010 

## Fix ANR when navigating up to initial page

I confirm that the crash does not occur. I verified `Go Up` button and `arrow icon`.

I noticed one unexpected behavior.
1. User starts `repeat_group_form.xml` form
2. User does not add repeat groups
3. User opens `Jump Screen`
4. User clicks on `Go Up` button or arrow icon.
5. User is moved to view with only `Group Name`
6. User changes device orientation and sees list of question in `Group Name`

| | |
|---|---|
|![screenshot_20190129-102841](https://user-images.githubusercontent.com/3907554/51898263-d4a04400-23b0-11e9-977c-b54cf384d89f.png)|![screenshot_20190129-102845](https://user-images.githubusercontent.com/3907554/51898275-dc5fe880-23b0-11e9-8d79-d461b2ccd35f.png)|

Second, related case:
1. User starts `repeat_group_form.xml` form
2. User does not add repeat groups
3. User opens `Jump Screen`
4. User clicks on `Go to End` button
5. User sees `Save Form and Exit` button
6. User opens `Jump Screen`

User sees a list of question in `Group Name` but I think that expected is view with only `Group Name`


## Hide groups from repeat picker

Tested with success

Verified on Android 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.1.
Verified form `repeat_group_form.xml`

## Fix bin icon not showing up for some repeat instances

Tested with success
Tested on Android: 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.1

Verified forms:
* 1648
* fewRepeatGroups
* nested-index-repeat
* nested- repeats
* nested-repeats-complex
* nRepeats
* nRepeatsWithoutInstanceID
* one repeat
* Repeat Group
* RepeatGroupAndGroup
* RepeatGroupAndGroup2
* RepeatGroupNew
* RepeatSimple
* repeatTest
* TestRepeat

cooperka committed:
Remove redundant 'go up' button from bottom

cooperka committed:
Also prevent *entering* single group from beginning

Otherwise forms with 1 top-level group will be entered automatically,
and that's not expected.

cooperka commented:
@mmarciniak90 originally if there was only a single group on the screen it was entered automatically (or exited automatically when navigating up), but that behavior doesn't make sense on the very first screen, so I updated it.

> 6. User changes device orientation and sees list of question in Group Name

Fixed.

> 6. User opens Jump Screen
>
> User sees a list of question in Group Name but I think that expected is view with only Group Name

Fixed.

mmarciniak90 commented:
@yanokwa @cooperka @grzesiek2010 

I have one suggestion to `Up` icon related to https://github.com/opendatakit/collect/issues/2831

This icon is visible always when I have groups. Adding and deleting depend on kind of group. 
As a user, I would prefer to have `Up` icon always at the same place, as the first icon from the right side.
This change prevents unintentional adding/deleting group.

| | | |
|---|---|---|
|![screenshot_20190129-214551](https://user-images.githubusercontent.com/3907554/51922667-c4588b00-23e9-11e9-9a4f-e98ea4bf3668.png)|![screenshot_20190129-214600](https://user-images.githubusercontent.com/3907554/51922665-c3bff480-23e9-11e9-8ade-4b40e4627ae2.png)|![screenshot_20190129-214556](https://user-images.githubusercontent.com/3907554/51922666-c3bff480-23e9-11e9-80f1-b7d6945e7c72.png)|

mmarciniak90 commented:
@cooperka, I manipulated in your form and I'm not sure that I understand one case.

Why the `readonly` question is not visible in `Jump screen` for `Group Name` group?

[repeat_group_form_changed.xml.txt](https://github.com/opendatakit/collect/files/2808732/repeat_group_form_changed.xml.txt)


|Grp1|Group Name|Grp2222|
|---|---|---|
|![screenshot_20190129-223130](https://user-images.githubusercontent.com/3907554/51925787-1dc3b880-23f0-11e9-8f9c-8835450b7b38.png)|![screenshot_20190129-223141](https://user-images.githubusercontent.com/3907554/51925801-24523000-23f0-11e9-8420-e8979d18e1fb.png)|![screenshot_20190129-223152](https://user-images.githubusercontent.com/3907554/51925811-2916e400-23f0-11e9-9e14-fafcaece916b.png)|
|![screenshot_20190129-223203](https://user-images.githubusercontent.com/3907554/51925829-2f0cc500-23f0-11e9-98c9-58e371c9b2bd.png)|![screenshot_20190129-223210](https://user-images.githubusercontent.com/3907554/51925840-3207b580-23f0-11e9-9549-c9f908db9a4a.png)|![screenshot_20190129-223216](https://user-images.githubusercontent.com/3907554/51925850-359b3c80-23f0-11e9-9e20-2c3220561583.png)|

cooperka commented:
> Why the readonly question is not visible in Jump screen for Group Name group?

I think because the label is blank. This is old code that I did not change (see [FormHierarchyActivity#L601](https://github.com/opendatakit/collect/blame/master@%7B2019-01-28%7D/collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java#L601)):

```java
    // show the question if it is an editable field.
    // or if it is **read-only** and the label is **not blank**.
```

> I would prefer to have Up icon always at the same place, as the first icon from the right side.

I agree. Originally we wanted the "up" icon to be second to line up with the "open jump menu" button in forms (and add/delete was always there so all icons always stayed in the same spot), but now with visible **groups** there isn't always an add/delete button.

I will move "up" to always be first on the right.

cooperka committed:
Move 'go up' button to always far right

grzesiek2010 commented:
Thanks for the explanation @cooperka It seems strange to me as well so I filled an issue where we can discuss it https://github.com/opendatakit/collect/issues/2836

mmarciniak90 commented:
@cooperka thanks for the explanation.

**Tested with success**

Verified on Android: 4.2, 4.4, 5.1, 6.0, 7.0, 8.1

Verified cases:
*  `up` icon is on the right
* Fix ANR when navigating up to initial page
* Hide groups from repeat picker
* Fix bin icon not showing up for some repeat instances

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"
